### DISCUSSION TITLE
Republish Rules Action via Environment Variables

### DISCUSSION THREAD

**Discussion by BeAddicted** [emqx/emqx#14285] [Q&A] [answered]
I am having trouble configuring the rules engine with environment variables (or config file). It seems like I can't find any documentation for this.

I found the documentation to setup a source like this:

```
      EMQX_SOURCES__MQTT__TEST__PARAMETERS__TOPIC: 'remoteTest'
      EMQX_SOURCES__MQTT__TEST__CONNECTOR: 'datacenter'
```
And I found out how to configure a rule with this source (even though I could not find any documentation on this)

`      EMQX_RULE_ENGINE__RULES__TEST__SQL: 'SELECT  *  FROM  "$$bridges/mqtt:test"'`

I also found out that I can set actions for rules with:

`      EMQX_RULE_ENGINE__RULES__TEST__ACTIONS: ["type:name"]`

But I don't know how to create a republish action (since also they are usually not named when creating them over the dashboard and there is no documentation for this.)

Is there any documentation on how to do this via Env? (or config file since the parameters there are equivalent)

**Comment by id** [comment, marked as answer]
@BeAddicted the simplest way is usually to configure what you need in UI, then run `emqx ctl conf show <root key>', and translate output to environment variables.

I'm not sure where you got your example, but those variables will not work with 5.x version.

I just created a simple republish action. Here is how `emqx ctl` output looks like:

```
$ emqx ctl conf show rule_engine
rule_engine {
  ignore_sys_message = true
  jq_function_default_timeout = "10s"
  rules {
    rule_republish {
      actions = [
        {
          args {
            direct_dispatch = false
            mqtt_properties {}
            payload = ""
            qos = 0
            retain = false
            topic = "t/2"
            user_properties = ""
          }
          function = republish
        }
      ]
      description = ""
      enable = true
      metadata {created_at = 1732639346251}
      name = ""
      sql = """~
        SELECT
          *
        FROM
          "t/1"~"""
    }
  }
}
```

The rules for environment variables are simple:
* They start with prefix `EMQX_`
* Each next level is separated with double underscore `__`.
* If config item is a list, each element is referenced via its index starting from 1, for example `EMQX_rootkey__listname__1__listitem`.

```
EMQX_rule_engine__rules__rule_republish__enable=true
EMQX_rule_engine__rules__rule_republish__sql='SELECT * FROM "t/1"'
EMQX_rule_engine__rules__rule_republish__actions__1__args__direct_dispatch=false
EMQX_rule_engine__rules__rule_republish__actions__1__args__payload=""
EMQX_rule_engine__rules__rule_republish__actions__1__args__qos=0
EMQX_rule_engine__rules__rule_republish__actions__1__args__retain=false
EMQX_rule_engine__rules__rule_republish__actions__1__args__topic=t/2
EMQX_rule_engine__rules__rule_republish__actions__1__function=republish
```

**Reply by BeAddicted** [reply]
Thank you! Got it working like this.
